### PR TITLE
Improve dataframe sorting in KTNB UI by handling non-comparable columns

### PR DESCRIPTION
**Pull request by ermolenkodev**
Fixes [#822](https://github.com/Kotlin/dataframe/issues/822)

### CODE REVIEW

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/jupyter/KotlinNotebookPluginUtils.kt [hunk 1]
```diff
@@ -79,27 +88,80 @@ public object KotlinNotebookPluginUtils {
...
+        val secondValue = (row2.getValueOrNull(key)?.toString() ?: "")
+
+        return if (desc) {
+            secondValue.compareTo(firstValue)
+        } else {
+            firstValue.compareTo(secondValue)
+        }
+    }
+
+    private fun KType.isComparable(): Boolean = this.isSubtypeOf(typeOf<Comparable<*>>())
```
Comment (Jolanrensen) [inline core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/jupyter/KotlinNotebookPluginUtils.kt:164]: This check breaks whenever you have, say, String and Int in a single column. Yes, they are both `Comparable`, but not comparable to each other. They are `Comparable<Nothing>` and calling the `compare` there will break.

Check out the fix I made here for `AnyCol.isComparable()`: https://github.com/Kotlin/dataframe/pull/726/files#diff-68de50cc685a57fef49b535d54b8dc925caecc39ba0bc1de3d312ffeb25e8a60

In this case, if we get `Comparable<Nothing>` we should probably just compare them as `String` again, so returning `false` here should be fine (I think).

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/jupyter/KotlinNotebookPluginUtils.kt [hunk 2]
```diff
@@ -79,27 +88,80 @@ public object KotlinNotebookPluginUtils {
...
 
-            val sortKeys = columnPaths.map { path ->
-                ColumnPath(path)
+        return df.sortWith(comparator)
+    }
+
+    private fun createComparator(sortKeys: List<ColumnPath>, isDesc: List<Boolean>): Comparator<DataRow<*>> {
+        return Comparator { row1, row2 ->
+            for ((key, desc) in sortKeys.zip(isDesc)) {
+                val comparisonResult = if (row1.df().getColumn(key).type.isComparable()) {
```
Comment (Jolanrensen) [inline core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/jupyter/KotlinNotebookPluginUtils.kt:116]: actually, we already have an `AnyCol.isComparable()` function :) it includes the fix I describe in the other comment

**Review by Jolanrensen — CHANGES_REQUESTED** [review]
Nicely done! A good number of tests too :). Aside from my note about isComparable it looks good to me

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/jupyter/KotlinNotebookPluginUtils.kt (hunk 2 above)
Comment (ermolenkodev) [reply core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/jupyter/KotlinNotebookPluginUtils.kt]: Thank you for noticing the problem!

**Review by Jolanrensen — APPROVED** [review]
Thanks for the extra test :)

**Pull request merged**